### DISCUSSION TITLE
Siemens Skyra coil configuration

### DISCUSSION THREAD

**Discussion by eozh** [shimming-toolbox/shimming-toolbox#475] [Q&A] [answered]
I have tried using the Shimming Toolbox with a Siemens Skyra scanner. When running B0 shim, I get 
`WARNING:shimmingtoolbox.coils.coil:Manufacturer Skyra not implemented, bounds might not be respected. Setting initial shim_setting to 0`
After searching the source code, it looks like only Prisma Fit scanner is supported. Would you expect the calculated shim offsets be valid for the Skyra scanner? Do I need to modify the coil_config.json file and enter the limits from the Manual Adjustments screen? Or would you recommend full coil profile calibration? Thank you!

**Comment by po09i** [comment]
Hi @eozh!

Here is some context: We try to output the coefficients with the same units as the ones in the adjustment window (if you use '--output-file-format-scanner' with either 'slicewise-ch', 'slicewise-coil', 'chronological-ch', 'chronological-coil') so that it is easy to change at the scanner. We have validated the Prisma Fit and we log this warning because other scanners have not been validated. Here are the info for the Prisma Fit:
- order 0 units: Hz
- order 1 units: mT/m
- order 2 units: mT/m^2
- Ordering: F0, X, Y, Z, Z2, ZX, ZY, X2 - Y2, XY
- Bounds: 
        [123100100, 123265000]
        [-2300, 2300],
        [-2300, 2300],
        [-2300, 2300],
        [-4959.01, 4959.01],
        [-3551.29, 3551.29],
        [-3503.299, 3503.299],
        [-3551.29, 3551.29],
        [-3487.302, 3487.302]

I have not seen anything that would suggest that the 1st order units, 2nd order units, ordering are different for different Siemens scanners, but I could be wrong. What is probably different are the bounds (the maximum and minimum values that the f0/gradient/shim coil can achieve).

> Would you expect the calculated shim offsets be valid for the Skyra scanner?

If the units for the 1st order and 2nd order are the same for the Skyra as the Prisma, and that the larmor frequency of the Skyra is within the limits of the coil_config.json file (change it manually if it is not the case), then I expect that the calculated shim offsets would be correct for the Skyra as well. However, I suggest you test it out. We typically validate with a strong gradient along one direction and make sure that the output of the toolbox is correct according to the strength and direction chosen.

> Do I need to modify the coil_config.json file and enter the limits from the Manual Adjustments screen?

Yes so that the limits of the scanner are respected.

We would love to know your findings as well as the 0th, 1st and 2nd order limits of the Skyra so that we can update the toolbox so that it is automated in the future.

Let me know if you have any other questions,

Alex

**Reply by eozh** [reply]
Thank you, Alex. I have created a config file with the limits for the Skyra scanner (added .txt for github to accept it): [skyra.json.txt](https://github.com/shimming-toolbox/shimming-toolbox/files/12718403/skyra.json.txt)

I've also ran a couple of tests:
_I used a cylindrical phantom in a 20-channel head and neck coil. For the field mapping sequence I used gre_brain double echo sequence from the Siemens tree, adjusted to increase resolution to 96x96x96 pixels. In the shimming toolbox I used "volume" mode (not per-slice), 1st order shim and mask based on a threshold. Using a "box" mask that extended beyond the phantom gave me bad results, probably due to empty voxels within the mask._

**1.** For the first test I performed a Siemens 3D shim with GRE Brain and then acquired the field map. When I used the shimming toolbox, it gave me very small shim offsets (Freq: -1.05, X:0.7, Y: 0.33, Z: 0.67), as expected. The FWHM in the Manual Shim dialog was 45.3.

**2.** For the second test, I artificially created a bad shim by adding offsets to shim values (X: -100, Y: +100, Z: +100) and acquired the field map. I expected the shimming toolbox to give me the opposite offsets (+100, -100, -100). Instead the output was (Freq: -28.33, X: 95.67, Y: -81.83, Z: -95.14). This was somewhat similar to what I expected, but a bit off. FWHM in the Manual Shim dialog was 101 (compared to 45.3 in the first test with Siemens 3D shim).

Do you have any thoughts on what may be causing this discrepancy? It seems to me that there may be some scaling factor missing between the shimming toolbox output and the values in the Adjustment dialog. Do you know if we can do anything to calibrate the gradient strength for the shimming toolbox? Or should we reach out to Siemens and ask them if they do any calibration in their pipeline?

Eugene

**Reply by po09i** [reply, marked as answer]
> probably due to empty voxels within the mask.

Indeed, you would want the mask for the B0 shim to only include voxels that have been calculated by the field map.

> Do you have any thoughts on what may be causing this discrepancy?

Shimming Toolbox minimizes the field map in the given ROI (your threshold mask) using different algorithms (pseudo inverse, SLSQP, quadprog) and criteria (MSE, MAE). Given that the pipelines and algorithms are not the same, the toolbox will most likely give different results. I expect them to be worse in the manual 3D shim window for a global shim since I believe (at least for the f0 optimization), Siemens optimizes for the smallest line-width. However, that is not necessarily a bad thing (e.g.: the ROI was more specific than Siemens' shim box) but it could indeed be a problem since you mentioned you used a threshold so most likely most of the volume.

There are other things that could lead to differences between Siemens' results and Shimming Toolbox's results.
- A field-map that does not accurately reflect the true field (eddy currents, etc)
- Theoretical vs experimental gradients/f0 (as you mentioned)

> Do you know if we can do anything to calibrate the gradient strength for the shimming toolbox?

There is currently no straightforward way for the user to modify the theoretical shims. However one way to do so would be to treat the scanner as if they were a "custom coil". This would involve acquiring field-maps at different shim currents (e.g.: -100 uT/m, 100 uT/m) for all channels of the shim system. Shimming Toolbox could then process those and create a "custom coil" that would be specific to your scanner. We do have a tutorial for creating custom coils [here](https://shimming-toolbox.org/en/latest/user_section/tutorials/create_b0_coil_profiles.html) if you choose that route.

I do know that GE and Philips provide some information about the non-linearity of their gradients/shim system. If Siemens offers a similar feature, I would be willing to implement a user-friendly way to update the theoretical shims with these values.

The first thing I would try however is trying out the "pseudo_inverse" optimizer (default is least_squares or quadprog depending on your shimming Toolbox version). Note that this is the most basic optimizer (it might not respect the constraints) but will typically offer the best performance and "should" rule out if the problem is with the field-map or if we should check for gradient non-linearities.

Alex

**Reply by eozh** [reply]
Hi Alex, I did try both "least squares" and "pseudo inverse" optimizers. I can try creating coil profile for the Skyra gradients and share the results with you.

Eugene

**Reply by po09i** [reply]
Great! Looking forward to the results!
Alex
